auth_brute_force/tests/test_brute_force.py

```diff
@@ -0,0 +1,361 @@
+        # TODO Remvove when fixed
+        try:
+            del current_thread().environ
+        except AttributeError:
+            pass
+        # Complex password to avoid conflicts with `password_security`
+        self.good_password = "Admin$%02584"
+        self.data_demo = {
+            "login": "demo",
+            "password": "Demo%&/(908409**",
+        }
+        with self.cursor() as cr:
+            env = self.env(cr)
+            env["ir.config_parameter"].set_param(
+                "auth_brute_force.max_by_ip_user", 3)
+            env["ir.config_parameter"].set_param(
+                "auth_brute_force.max_by_ip", 4)
+            # Clean attempts to be able to count in tests
+            env["res.authentication.attempt"].search([]).unlink()
+            # Make sure involved users have good passwords
+            env.user.password = self.good_password
+            env["res.users"].search([
+                ("login", "=", self.data_demo["login"]),
+            ]).password = self.data_demo["password"]
+
+    @skipUnless(can_import("odoo.addons.web"), "Needs web addon")
+    @mute_logger(*GARBAGE_LOGGERS)
+    def test_web_login_existing(self, *args):
+        """Remote is banned with real user on web login form."""
+        data1 = {
+            "login": "admin",
+            "password": "1234",  # Wrong
+        }
+        # Make sure user is logged out
+        self.url_open("/web/session/logout", timeout=30)
+        # Fail 3 times
+        for n in range(3):
+            response = self.url_open("/web/login", bytes(urlencode(data1)), 30)
+            # If you fail, you get /web/login again
+            self.assertTrue(
+                response.geturl().endswith("/web/login"),
+                "Unexpected URL %s" % response.geturl(),
+            )
+        # Admin banned, demo not
+        with self.cursor() as cr:
+            env = self.env(cr)
+            self.assertFalse(
+                env["res.authentication.attempt"]._trusted(
+                    "127.0.0.1",
+                    data1["login"],
+                ),
+            )
+            self.assertTrue(
+                env["res.authentication.attempt"]._trusted(
+                    "127.0.0.1",
+                    "demo",
+                ),
+            )
+        # Now I know the password, but login is rejected too
+        data1["password"] = self.good_password
+        response = self.url_open("/web/login", bytes(urlencode(data1)), 30)
+        self.assertTrue(
+            response.geturl().endswith("/web/login"),
+            "Unexpected URL %s" % response.geturl(),
+        )
+        # IP has been banned, demo user cannot login
+        with self.cursor() as cr:
+            env = self.env(cr)
+            self.assertFalse(
+                env["res.authentication.attempt"]._trusted(
+                    "127.0.0.1",
+                    "demo",
+                ),
+            )
+        # Attempts recorded
+        with self.cursor() as cr:
```

Comment:
Readabillty: These test cases seem quite repetitive, maybe it would be possible to factor out the common parts to reduce code repetition?

Reply:
Possibly, but the little details make it quite difficult to refactor, and after all repeating code in test cases is not so bad as in other parts of the code.